### DISCUSSION TITLE
Bun Shell - exit code and `await for (const chunk of $`…`.chunks())`

### DISCUSSION THREAD

**Discussion by TomasHubelbauer** [oven-sh/bun#8324] [Q&A]
Hi! Thanks for releasing Bun Shell!

I looked at the types and I only found this way to extract the exit code while also using one of the helpers, like `.text()`:

```typescript
import { $ } from 'bun';
import { expect } from 'bun:test';

const process = $`bun -e ${'const d = +performance.now().toString().slice(-2); console.log(d); process.exit(d);'}`;
const text = await process.text();
const { exitCode } = await process;
expect(exitCode).toEqual(+text);
```

I verify this doesn't run the command two separate times by checking the exit code and the output match the same `performance.now` instant slice.

Is this an okay way to get the exit code while also using one of the helpers, like `.text`?

When I did `await process.exitCode` first and then `await process.text()` I got an error:

> Shell is already running

Also, I think the helpers are buffered, right? I was curious about this in regard to the `.lines()` helper:

```typescript
import { $ } from 'bun';

for await (const line of $`bun -e "while (true) { console.log(Date.now()); await Bun.sleep(1000); }"`.lines()) {
  console.log({ line });
}
```

This code waits forever. Is there a way `.lines()` could have an overload or some other way so that the above snippet would work and consume the lines as they are printed to the standard output?

And related to that, if that is possible and will come, can we also get `.chunks` alongside it, so that we could consume the chunks as they are flushed to the buffer instead of the `.lines()` helper doing the line splitting for us?

```typescript
for await (const chunk of $`my-process`.chunks()) {
  // Process the flushed chunk here
}
```

Thanks!

@zackradisic you're probably the best person to ask this as the shell PR author I believe.

**Comment by zackradisic** [comment]
> Is this an okay way to get the exit code while also using one of the helpers, like .text?

I think we should make helpers like `.text()` and `.line()` be on the returned `ShellOutput` so the `ShellPromise` doesn't have to be awaited again

> This code waits forever. Is there a way .lines() could have an overload or some other way so that the above snippet would work and consume the lines as they are printed to the standard output?

Yep, at some point we will allow stdout/stderr to be consumed as ReadableStreams

**Reply by M-Gonzalo** [reply]
Yes please, especially for binary content, something like `.chunks(512)`. Anything that allows us to iterate over `stdout`
